File: phishdetect/features.py
"""URL feature extraction — 20 deterministic, string-only features.

Every feature is computed purely from the URL *string* (after the shared
deterministic parse in :mod:`phishdetect.urlparse`). No DNS, no HTTP, no
network of any kind. That is what lets the identical logic run client-side in
the browser (`web/src/features.ts` is a faithful port of this module).

The 20 features, in fixed order (see :data:`FEATURE_NAMES`)
-----------------------------------------------------------
=== ============================ ===================================================
#   name                         definition
=== ============================ ===================================================
1   url_length                   length of the input URL string (pre-scheme-synthesis)
2   hostname_length              length of the parsed host
3   path_length                  length of the path component (incl. leading ``/``)
4   num_dots                     count of ``.`` in the whole URL string
5   num_hyphens                  count of ``-`` in the hostname
6   num_subdomains               host labels in excess of the registrable domain
7   has_ip_host                  1 if the host is an IPv4/IPv6 literal, else 0
8   has_at_symbol                1 if the URL string contains ``@``, else 0
9   num_query_params             number of ``&``-separated query parameters
10  has_punycode                 1 if any host label starts ``xn--``, else 0
11  has_homograph                1 if the host has any non-ASCII character, else 0
12  is_https                     1 if the scheme is ``https``, else 0
13  num_digits_in_host           count of digits ``0-9`` in the host
14  digit_ratio_host             num_digits_in_host / hostname_length (0 if host empty)
15  suspicious_tld               1 if the eTLD is in :data:`SUSPICIOUS_TLDS`, else 0
16  is_shortener                 1 if the host is a known URL shortener, else 0
17  num_suspicious_keywords      count of phishing keywords found in the URL (lowercased)
18  has_double_slash_in_path     1 if ``//`` appears inside the path, else 0
19  num_special_chars            count of ``% = ? & _`` in the whole URL string
20  tld_length                   length of the eTLD string (e.g. ``co.uk`` -> 5)
=== ============================ ===================================================

The function :func:`extract_features` returns an ordered ``dict`` whose keys
are exactly :data:`FEATURE_NAMES`; :func:`feature_vector` returns the values as
a plain ``list[float]`` in that order (the form the model consumes).
"""

from __future__ import annotations

from .suffix_list import KNOWN_TLDS, MULTI_LABEL_SUFFIXES
from .urlparse import ParsedURL, parse_url

# --- Fixed, ordered feature list ------------------------------------------------
# The model's coefficients are aligned to THIS order. Never reorder; only append.
FEATURE_NAMES: list[str] = [
    "url_length",
    "hostname_length",
    "path_length",
    "num_dots",
    "num_hyphens",
    "num_subdomains",
    "has_ip_host",
    "has_at_symbol",
    "num_query_params",
    "has_punycode",
    "has_homograph",
    "is_https",
    "num_digits_in_host",
    "digit_ratio_host",
    "suspicious_tld",
    "is_shortener",
    "num_suspicious_keywords",
    "has_double_slash_in_path",
    "num_special_chars",
    "tld_length",
]

# Top-level domains disproportionately abused for phishing / malware. Mirrored
# verbatim in the TypeScript port.
SUSPICIOUS_TLDS: frozenset[str] = frozenset(
    {
        "zip",
        "mov",
        "xyz",
        "top",
        "tk",
        "gq",
        "ml",
        "cf",
        "ga",
        "country",
        "kim",
        "work",
        "click",
        "link",
        "gdn",
        "loan",
        "download",
        "review",
        "stream",
        "racing",
        "win",
        "bid",
        "party",
        "date",
        "men",
        "cricket",
        "science",
        "accountant",
        "trade",
    }
)

# Well-known URL-shortening services. A shortener hides the true destination,
# which is a phishing-relevant signal in its own right. Mirrored in the port.
URL_SHORTENERS: frozenset[str] = frozenset(
    {
        "bit.ly",
        "tinyurl.com",
        "goo.gl",
        "t.co",
        "ow.ly",
        "is.gd",
        "buff.ly",
        "cutt.ly",
        "rebrand.ly",
        "bit.do",
        "shorturl.at",
        "rb.gy",
        "tiny.cc",
        "t.ly",
        "soo.gd",
        "s.id",
        "lnkd.in",
        "db.tt",
        "qr.ae",
        "adf.ly",
        "tr.im",
        "v.gd",
    }
)

# Keywords commonly used in phishing URLs to imitate a trusted action/brand.
# Counted as the number of these substrings present in the lowercased URL.
SUSPICIOUS_KEYWORDS: list[str] = [
    "login",
    "secure",
    "account",
    "verify",
    "update",
    "bank",
    "confirm",
    "signin",
    "password",
    "webscr",
    "ebayisapi",
]


def _is_ipv4(host: str) -> bool:
    """Return True if ``host`` is a dotted-decimal IPv4 literal."""
    parts = host.split(".")
    if len(parts) != 4:
        return False
    for part in parts:
        if part == "" or len(part) > 3:
            return False
        if not all("0" <= c <= "9" for c in part):
            return False
        if int(part) > 255:
            return False
    return True


def _is_ipv6(host: str) -> bool:
    """Return True if ``host`` is a bracketed IPv6 literal (``[...]``)."""
    if not (host.startswith("[") and host.endswith("]")):
        return False
    inner = host[1:-1]
    if inner == "":
        return False
    # Must look like hex groups separated by ':' (with optional '::' compression
    # and an optional embedded IPv4 tail). A light structural check is enough.
    if "::" in inner:
        if inner.count("::") > 1:
            return False
    allowed = "0123456789abcdefABCDEF:."
    return all(c in allowed for c in inner) and ":" in inner


def _is_ip_host(host: str) -> bool:
    """Return True if the host is an IPv4 or IPv6 literal."""
    return _is_ipv4(host) or _is_ipv6(host)


def _host_is_ascii(host: str) -> bool:
    """Return True if every character of the host is ASCII (code point < 128)."""
    return all(ord(c) < 128 for c in host)


def split_registrable_domain(host: str) -> tuple[str, str, str]:
    """Split a host into ``(subdomains, registrable_domain, etld)``.

    Uses the bundled :data:`MULTI_LABEL_SUFFIXES` / :data:`KNOWN_TLDS`. The
    registrable domain is the eTLD plus exactly one label to its left ("eTLD+1").

    Examples
    --------
    ``www.example.co.uk`` -> ``("www", "example.co.uk", "co.uk")``
    ``mail.google.com``   -> ``("mail", "google.com", "com")``
    ``example.com``       -> ``("", "example.com", "com")``

    For an IP-literal host or a single-label host there is no registrable
    domain; the eTLD is returned empty.
    """
    if host == "" or _is_ip_host(host):
        return "", host, ""

    labels = host.split(".")
    if len(labels) < 2:
        # A bare single label has no public suffix split.
        return "", host, ""

    # Prefer the longest matching multi-label suffix.
    etld = ""
    if len(labels) >= 3:
        two = ".".join(labels[-2:])
        if two in MULTI_LABEL_SUFFIXES:
            etld = two
    if etld == "":
        last = labels[-1]
        # Whether or not ``last`` is a recognised TLD, treat it as the eTLD:
        # unknown new gTLDs still behave as single-label suffixes.
        etld = last
        _ = KNOWN_TLDS  # referenced for documentation/validation parity

    etld_labels = etld.count(".") + 1
    if len(labels) <= etld_labels:
        # Host *is* the suffix (rare, e.g. literally "co.uk") — no domain part.
        return "", host, etld

    registrable = ".".join(labels[-(etld_labels + 1) :])
    subdomain_labels = labels[: len(labels) - (etld_labels + 1)]
    return ".".join(subdomain_labels), registrable, etld


def _count_query_params(query: str) -> int:
    """Return the number of parameters in a raw query string.

    Parameters are the non-empty pieces obtained by splitting on ``&`` (and
    also ``;``, a legacy separator). An empty query has zero parameters.
    """
    if query == "":
        return 0
    count = 0
    for chunk in query.replace(";", "&").split("&"):
        if chunk != "":
            count += 1
    return count


def extract_features_from_parsed(parsed: ParsedURL) -> dict[str, float]:
    """Compute the 20-feature dict from an already-:func:`parse_url` result."""
    url_str = parsed.original
    url_lower = url_str.lower()
    host = parsed.host

    host_labels = host.split(".") if host else []
    _subdomains, _registrable, etld = split_registrable_domain(host)
    num_subdomains = len(_subdomains.split(".")) if _subdomains != "" else 0

    digits_in_host = sum(1 for c in host if "0" <= c <= "9")
    hostname_length = len(host)
    digit_ratio = (digits_in_host / hostname_length) if hostname_length > 0 else 0.0

    has_punycode = 1.0 if any(lbl.startswith("xn--") for lbl in host_labels) else 0.0
    has_homograph = 0.0 if _host_is_ascii(host) else 1.0

    num_keywords = sum(1 for kw in SUSPICIOUS_KEYWORDS if kw in url_lower)
    num_special = sum(1 for c in url_str if c in "%=?&_")

    features: dict[str, float] = {
        "url_length": float(len(url_str)),
        "hostname_length": float(hostname_length),
        "path_length": float(len(parsed.path)),
        "num_dots": float(url_str.count(".")),
        "num_hyphens": float(host.count("-")),
        "num_subdomains": float(num_subdomains),
        "has_ip_host": 1.0 if _is_ip_host(host) else 0.0,
        "has_at_symbol": 1.0 if "@" in url_str else 0.0,
        "num_query_params": float(_count_query_params(parsed.query)),
        "has_punycode": has_punycode,
        "has_homograph": has_homograph,
        "is_https": 1.0 if parsed.scheme == "https" else 0.0,
        "num_digits_in_host": float(digits_in_host),
        "digit_ratio_host": digit_ratio,
        "suspicious_tld": 1.0 if etld in SUSPICIOUS_TLDS else 0.0,
        "is_shortener": 1.0 if host in URL_SHORTENERS else 0.0,
        "num_suspicious_keywords": float(num_keywords),
        "has_double_slash_in_path": 1.0 if "//" in parsed.path else 0.0,
        "num_special_chars": float(num_special),
        "tld_length": float(len(etld)),
    }
    # Guard: dict order must equal FEATURE_NAMES order.
    assert list(features.keys()) == FEATURE_NAMES
    return features


def extract_features(url: str) -> dict[str, float]:
    """Extract the 20-feature dict from a raw URL string.

    The returned dict's keys are exactly :data:`FEATURE_NAMES`, in order.
    """
    return extract_features_from_parsed(parse_url(url))


def feature_vector(url: str) -> list[float]:
    """Return the 20 feature values as an ordered ``list[float]``.

    This is the exact form the logistic-regression model consumes.
    """
    feats = extract_features(url)
    return [feats[name] for name in FEATURE_NAMES]

File: web/src/features.ts
/**
 * URL feature extraction — TypeScript port of `phishdetect/features.py`.
 *
 * Extracts the SAME 20 string-only features the Python library does, so the
 * browser and the CLI feed the model identical vectors. No network of any kind.
 *
 * Parity notes
 * ------------
 * - Python's `len(str)`, `str.count(...)` and `ord(c)` operate on Unicode code
 *   *points*. JavaScript strings are UTF-16; naive `.length` and indexing count
 *   *code units*, which differ for astral characters. Every length/count here
 *   therefore iterates with `[...str]` / `for...of`, which yield code points,
 *   matching Python exactly.
 * - The feature order is fixed and identical to `FEATURE_NAMES` in Python.
 *
 * The shared `fixtures/parity_urls.json` fails if this drifts from `features.py`.
 */

import { KNOWN_TLDS, MULTI_LABEL_SUFFIXES } from './suffixList'
import { parseUrl, type ParsedURL } from './urlparse'

/** Fixed, ordered feature list — the model's coefficients align to this order. */
export const FEATURE_NAMES = [
  'url_length',
  'hostname_length',
  'path_length',
  'num_dots',
  'num_hyphens',
  'num_subdomains',
  'has_ip_host',
  'has_at_symbol',
  'num_query_params',
  'has_punycode',
  'has_homograph',
  'is_https',
  'num_digits_in_host',
  'digit_ratio_host',
  'suspicious_tld',
  'is_shortener',
  'num_suspicious_keywords',
  'has_double_slash_in_path',
  'num_special_chars',
  'tld_length',
] as const

/** A union of the 20 feature names. */
export type FeatureName = (typeof FEATURE_NAMES)[number]

/** The 20-feature record produced by {@link extractFeatures}. */
export type FeatureVector = Record<FeatureName, number>

/** TLDs disproportionately abused for phishing. Mirrors SUSPICIOUS_TLDS. */
export const SUSPICIOUS_TLDS: ReadonlySet<string> = new Set<string>([
  'zip',
  'mov',
  'xyz',
  'top',
  'tk',
  'gq',
  'ml',
  'cf',
  'ga',
  'country',
  'kim',
  'work',
  'click',
  'link',
  'gdn',
  'loan',
  'download',
  'review',
  'stream',
  'racing',
  'win',
  'bid',
  'party',
  'date',
  'men',
  'cricket',
  'science',
  'accountant',
  'trade',
])

/** Well-known URL-shortening services. Mirrors URL_SHORTENERS. */
export const URL_SHORTENERS: ReadonlySet<string> = new Set<string>([
  'bit.ly',
  'tinyurl.com',
  'goo.gl',
  't.co',
  'ow.ly',
  'is.gd',
  'buff.ly',
  'cutt.ly',
  'rebrand.ly',
  'bit.do',
  'shorturl.at',
  'rb.gy',
  'tiny.cc',
  't.ly',
  'soo.gd',
  's.id',
  'lnkd.in',
  'db.tt',
  'qr.ae',
  'adf.ly',
  'tr.im',
  'v.gd',
])

/** Phishing keywords counted in the lowercased URL. Mirrors SUSPICIOUS_KEYWORDS. */
export const SUSPICIOUS_KEYWORDS: readonly string[] = [
  'login',
  'secure',
  'account',
  'verify',
  'update',
  'bank',
  'confirm',
  'signin',
  'password',
  'webscr',
  'ebayisapi',
]

/** Count occurrences of `ch` in `text` (single-character code-point counting). */
function countChar(text: string, ch: string): number {
  let n = 0
  for (const c of text) {
    if (c === ch) {
      n += 1
    }
  }
  return n
}

/** Count code points in `text` (Python `len(str)` semantics). */
function codePointLength(text: string): number {
  let n = 0
  for (const _c of text) {
    n += 1
  }
  return n
}

/** Return true if `host` is a dotted-decimal IPv4 literal. */
function isIpv4(host: string): boolean {
  const parts = host.split('.')
  if (parts.length !== 4) {
    return false
  }
  for (const part of parts) {
    if (part === '' || part.length > 3) {
      return false
    }
    if (![...part].every((c) => c >= '0' && c <= '9')) {
      return false
    }
    if (Number.parseInt(part, 10) > 255) {
      return false
    }
  }
  return true
}

/** Return true if `host` is a bracketed IPv6 literal (`[...]`). */
function isIpv6(host: string): boolean {
  if (!(host.startsWith('[') && host.endsWith(']'))) {
    return false
  }
  const inner = host.slice(1, -1)
  if (inner === '') {
    return false
  }
  if (inner.includes('::')) {
    if ((inner.match(/::/g) ?? []).length > 1) {
      return false
    }
  }
  const allowed = '0123456789abcdefABCDEF:.'
  return [...inner].every((c) => allowed.includes(c)) && inner.includes(':')
}

/** Return true if the host is an IPv4 or IPv6 literal. */
function isIpHost(host: string): boolean {
  return isIpv4(host) || isIpv6(host)
}

/** Return true if every character of the host is ASCII (code point < 128). */
function hostIsAscii(host: string): boolean {
  for (const c of host) {
    if ((c.codePointAt(0) ?? 0) >= 128) {
      return false
    }
  }
  return true
}

/**
 * Split a host into `[subdomains, registrableDomain, etld]`.
 * Faithful port of `split_registrable_domain` in Python.
 */
export function splitRegistrableDomain(host: string): [string, string, string] {
  if (host === '' || isIpHost(host)) {
    return ['', host, '']
  }
  const labels = host.split('.')
  if (labels.length < 2) {
    return ['', host, '']
  }

  let etld = ''
  if (labels.length >= 3) {
    const two = labels.slice(-2).join('.')
    if (MULTI_LABEL_SUFFIXES.has(two)) {
      etld = two
    }
  }
  if (etld === '') {
    etld = labels[labels.length - 1]
    void KNOWN_TLDS // referenced for parity with the Python module
  }

  const etldLabels = countChar(etld, '.') + 1
  if (labels.length <= etldLabels) {
    return ['', host, etld]
  }

  const registrable = labels.slice(-(etldLabels + 1)).join('.')
  const subdomainLabels = labels.slice(0, labels.length - (etldLabels + 1))
  return [subdomainLabels.join('.'), registrable, etld]
}

/** Count parameters in a raw query string (split on `&` and legacy `;`). */
function countQueryParams(query: string): number {
  if (query === '') {
    return 0
  }
  let count = 0
  for (const chunk of query.replace(/;/g, '&').split('&')) {
    if (chunk !== '') {
      count += 1
    }
  }
  return count
}

/** Compute the 20-feature record from an already-{@link parseUrl}d result. */
export function extractFeaturesFromParsed(parsed: ParsedURL): FeatureVector {
  const urlStr = parsed.original
  const urlLower = urlStr.toLowerCase()
  const host = parsed.host

  const hostLabels = host ? host.split('.') : []
  const [subdomains, , etld] = splitRegistrableDomain(host)
  const numSubdomains = subdomains !== '' ? subdomains.split('.').length : 0

  let digitsInHost = 0
  for (const c of host) {
    if (c >= '0' && c <= '9') {
      digitsInHost += 1
    }
  }
  const hostnameLength = codePointLength(host)
  const digitRatio = hostnameLength > 0 ? digitsInHost / hostnameLength : 0

  const hasPunycode = hostLabels.some((lbl) => lbl.startsWith('xn--')) ? 1 : 0
  const hasHomograph = hostIsAscii(host) ? 0 : 1

  let numKeywords = 0
  for (const kw of SUSPICIOUS_KEYWORDS) {
    if (urlLower.includes(kw)) {
      numKeywords += 1
    }
  }

  let numSpecial = 0
  for (const c of urlStr) {
    if ('%=?&_'.includes(c)) {
      numSpecial += 1
    }
  }

  const features: FeatureVector = {
    url_length: codePointLength(urlStr),
    hostname_length: hostnameLength,
    path_length: codePointLength(parsed.path),
    num_dots: countChar(urlStr, '.'),
    num_hyphens: countChar(host, '-'),
    num_subdomains: numSubdomains,
    has_ip_host: isIpHost(host) ? 1 : 0,
    has_at_symbol: urlStr.includes('@') ? 1 : 0,
    num_query_params: countQueryParams(parsed.query),
    has_punycode: hasPunycode,
    has_homograph: hasHomograph,
    is_https: parsed.scheme === 'https' ? 1 : 0,
    num_digits_in_host: digitsInHost,
    digit_ratio_host: digitRatio,
    suspicious_tld: SUSPICIOUS_TLDS.has(etld) ? 1 : 0,
    is_shortener: URL_SHORTENERS.has(host) ? 1 : 0,
    num_suspicious_keywords: numKeywords,
    has_double_slash_in_path: parsed.path.includes('//') ? 1 : 0,
    num_special_chars: numSpecial,
    tld_length: codePointLength(etld),
  }
  return features
}

/** Extract the 20-feature record from a raw URL string. */
export function extractFeatures(url: string): FeatureVector {
  return extractFeaturesFromParsed(parseUrl(url))
}

/** Return the 20 feature values as an ordered `number[]` (the model's input). */
export function featureVector(url: string): number[] {
  const feats = extractFeatures(url)
  return FEATURE_NAMES.map((name) => feats[name])
}
